### PR TITLE
Added API to extract raw response.

### PR DESCRIPTION
**Pull request by katmsft**
This is needed because in the convenience layer, some APIs will reconstruct the returned Response<T> object. Without the API the convenience layer have to make a copy of the existing RawResponse, which introduces unneeded memory copy.
E.g: https://github.com/Azure/azure-sdk-for-cpp/blob/master/sdk/storage/src/datalake/path_client.cpp#L290

### CODE REVIEW

**Comment by katmsft** [comment]
> `Response<T>` already provides a `GetRawResponse` function returning a non-const reference, so technically we can move from this reference. But I don't know if this is a good design.
> 
> Anyway, I'll just approve this PR and let Core team make the decision whether or not to accept this PR.

As talked offline, `GetRawResponse` is not consistent with `ExtractValue`, the latter clearly shows that the ownership is changed while the former does not, so I don't think it's a good practise to move the referenced value out by calling `GetRawResponse`.

**Comment by katmsft** [comment]
Note: Current Master's CI is broken, resulting in this PR's CI being broken.

**Comment by katmsft** [comment]
Note: This also fixes the issue where `Response<T>` cannot be moved passed as return value due to not having a move constructor.

File: sdk/core/azure-core/inc/response.hpp [hunk 1]
```diff
@@ -23,11 +23,27 @@ namespace Azure { namespace Core {
 
     T& operator=(const Response& other) = delete;
 
+    Response(Response&& other)
+        : m_value(std::move(other.m_value)), m_rawResponse(std::move(other.m_rawResponse))
+    {
+    }
+
+    T& operator=(const Response&& other)
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/response.hpp:31]: no const

File: sdk/core/azure-core/inc/response.hpp [hunk 2]
```diff
@@ -23,11 +23,27 @@ namespace Azure { namespace Core {
...
+    {
+      m_value = std::move(other.m_value);
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/response.hpp:33]:
```C++
if (&other != this)
```

File: sdk/core/azure-core/inc/response.hpp [hunk 3]
```diff
@@ -23,11 +23,27 @@ namespace Azure { namespace Core {
 
     T& operator=(const Response& other) = delete;
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/response.hpp:24]: Maybe also mark copy constructor deleted? Is it necessary?

**Comment by vhvb1989** [comment]
>This is needed because in the convenience layer, some APIs will reconstruct the returned Response object.

Why do we want to reconstruct the RawResponse?  I though we only construct <T> from a RawResponse.  Can you elaborate on it?

File: sdk/core/azure-core/inc/response.hpp [hunk 4]
```diff
@@ -21,13 +21,16 @@ namespace Azure { namespace Core {
...
-    T& operator=(const Response& other) = delete;
-
     const T* operator->() const { return &this->m_value; };
     T* operator->() { return &this->m_value; };
     T& operator*() { return this->m_value; };
     const T& operator*() const { return this->m_value; };
 
     T ExtractValue() { return std::move(this->m_value); }
+
+    std::unique_ptr<Http::RawResponse> ExtractRawResponse()
```
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/response.hpp:31]: I'm not sure that "Extract" is idiomatic to C++. I'd like @CaseyCarter / @BillyONeal to comment on pattern and what is the best naming.

**Comment by JinmingHu-MSFT** [comment]
@katmsft I guess i'm going to merge my PR https://github.com/Azure/azure-sdk-for-cpp/pull/327 first, where I already removed `T& operator=(const Response& other) = delete;`.

File: sdk/core/azure-core/inc/response.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/core/azure-core/inc/response.hpp:31]: Sure, this will be blocked by @CaseyCarter and @BillyONeal 's comment then.

BTW, if this is not idiomatic for C++, we need to also change the name of `ExtractValue`.

File: sdk/core/azure-core/inc/response.hpp [hunk 5]
```diff
@@ -21,13 +21,16 @@ namespace Azure { namespace Core {
...
     Http::RawResponse& GetRawResponse() { return *this->m_rawResponse; }
 
-    T& operator=(const Response& other) = delete;
-
     const T* operator->() const { return &this->m_value; };
     T* operator->() { return &this->m_value; };
     T& operator*() { return this->m_value; };
     const T& operator*() const { return this->m_value; };
 
     T ExtractValue() { return std::move(this->m_value); }
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/response.hpp:29]:
```suggestion
    T&& ExtractValue() { return std::move(this->m_value); }
```

File: sdk/core/azure-core/inc/response.hpp (hunk 4 above)
Comment (BillyONeal) [inline sdk/core/azure-core/inc/response.hpp:31]:
```suggestion
    std::unique_ptr<Http::RawResponse>&& ExtractRawResponse()
```

Comment (BillyONeal) [reply sdk/core/azure-core/inc/response.hpp:31]: I think Extract is fine.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/response.hpp:31]: I still believe we don't need `ExtracRawResponse`.  I would expect only T to be extracted at one point from responseT to forget about ResposeT (and rawResponse).  But do we expect to extract rawResponse and forget about T?

Comment (BillyONeal) [reply sdk/core/azure-core/inc/response.hpp:31]: This kinda stuff is why I was recommending literally:

```
template<class T>
struct Response {
    std::unique_ptr<RawResponse> RawResponse;
    T Value;
};
```

i.e. `Response` is just an aggregate with 2 members, akin to `std::pair`.

Alternately consider:

```
template<class T>
struct Response : T { // note derivation from T
    std::unique_ptr<RawResponse> RawResponse;
};
```

Comment (CaseyCarter) [reply sdk/core/azure-core/inc/response.hpp:31]: We have no real idioms<sup>1</sup> for "this function yanks something out of this object and returns it." I suggested `ExtractValue` because the original name `GetValue` suggests that it's returning a copy. "Extract" more clearly implies that the target object is modified by the call.

<sup>1</sup>The closest examples we have in the Standard Library are functions that require rvalue arguments and return rvalue references like
```c++
// "tuple of strings" means tuple<string, /* maybe other things */>; these return string&&
std::get<0>(std::move(some_tuple_of_strings));
std::get<0>(some_function_that_returns_a_tuple_of_strings());
```
or `optional::value()`, which can be called on `const`/non-`const` lvalue/rvalue `optional`s, which returns a `T&&` when called on a non-const rvalue `optional<T>`. There's no consensus even among WG21 whether such functions should return `T` (always extract the `T`) or `T&&` (give the caller the option to extract the `T`), or even whether it's a good idea to have `std::move(something).some_member()` do something other than leave `something` in a garbage state. 

TLDR: I don't consider the Standard Library to provide good precedent here.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/response.hpp:31]: > 
> 
> This kinda stuff is why I was recommending literally:
> 
> ```
> template<class T>
> struct Response {
>     std::unique_ptr<RawResponse> RawResponse;
>     T Value;
> };
> ```
> 
> i.e. `Response` is just an aggregate with 2 members, akin to `std::pair`.
> 
> Alternately consider:
> 
> ```
> template<class T>
> struct Response : T { // note derivation from T
>     std::unique_ptr<RawResponse> RawResponse;
> };
> ```

if we use struct on the template and expose the fields as public, is it possible to have them mutated?  
Also, I was looking for a way to make ResponseT to be use as a T directly, instead of doing `Response.value.method()` to just doing `Response->method()`

Comment (katmsft) [reply sdk/core/azure-core/inc/response.hpp:31]: > > This is needed because in the convenience layer, some APIs will reconstruct the returned Response object.
> 
> Why do we want to reconstruct the RawResponse? I though we only construct from a RawResponse. Can you elaborate on it?

Sorry, missed this comment. Basically the `Response<T>` is returned by the protocol and convenience layer APIs. Sometimes the returned value is again parsed or reconstructed to another `Response<T>`.

Two examples of the necessity:
1. Some DataLake APIs calls Blob service convenience APIs for interop as a universal strategy among track two SDKs. Blob APIs would return `Response<SomeBlobResult>` and the DataLake API needs to reconstruct it to `Response<SomeDataLakeResult>`.
2. Some protocol layer API has multiple functionalities, e.g. the set blob properties REST API can set the blob properties and metadata, so the protocol layer API `SetPropeties` will be called in `SetProperties` and `SetMetadata` in convenience layer, the `Response<PropertiesInfo>` returned from protocol layer will then need to be changed to `Response<MetadataInfo` for `SetMetadata` APIs.

**Comment by katmsft** [comment]
Validated and merged in https://github.com/Azure/azure-sdk-for-cpp/pull/308
